packages/vite-plugin-cloudflare/src/index.ts

```diff
@@ -419,7 +384,84 @@ export function cloudflare(pluginConfig: PluginConfig = {}): vite.Plugin[] {
 				}
 			},
 		},
+		// Plugin that can provide Node.js compatibility support for Vite Environments that are hosted in Cloudflare Workers.
+		{
+			name: "vite-plugin-cloudflare:nodejs-compat",
+			config() {
+				// Configure Vite with the Node.js polyfill aliases
+				// We have to do this across the whole Vite config because it is not possible to do it per Environment.
+				// These aliases are to virtual modules that then get Environment specific handling in the resolveId hook.
+				return {
+					resolve: {
+						alias: getNodeCompatAliases(),
+					},
+				};
+			},
+			configEnvironment(environmentName) {
+				// Ignore Node.js external modules when building environments that use Node.js compat.
+				const workerConfig = getWorkerConfig(environmentName);
+				if (isNodeCompat(workerConfig)) {
+					return {
+						build: {
+							rollupOptions: {
+								external: getNodeCompatExternals(),
+							},
+						},
+					};
+				}
+			},
+			async resolveId(source, importer, options) {
+				// Handle the virtual modules that come from Node.js compat aliases.
+				const from = maybeStripNodeJsVirtualPrefix(source);
+				if (!from) {
+					return;
+				}
+
+				const workerConfig = getWorkerConfig(this.environment.name);
+				if (!isNodeCompat(workerConfig)) {
+					return this.resolve(from, importer, options);
+				}
+
+				const unresolvedAlias = dealiasVirtualNodeJSImport(from);
+				const resolvedAlias = await this.resolve(
+					unresolvedAlias,
+					import.meta.url
+				);
+				assert(
+					resolvedAlias,
+					"Failed to resolve aliased nodejs import: " + unresolvedAlias
+				);
+
+				if (this.environment.mode === "dev" && this.environment.depsOptimizer) {
+					// Make sure the dependency optimizer is aware of this aliased import
+					this.environment.depsOptimizer.registerMissingImport(
+						unresolvedAlias,
+						resolvedAlias.id
+					);
+				}
+
+				return resolvedAlias;
+			},
+			async transform(code, id) {
+				// Inject the Node.js compat globals into the entry module for Node.js compat environments.
+				const workerConfig = getWorkerConfig(this.environment.name);
+				if (!isNodeCompat(workerConfig)) {
+					return;
+				}
+
+				const resolvedId = await this.resolve(workerConfig.main);
+				if (id === resolvedId?.id) {
+					return injectGlobalCode(id, code);
+				}
+			},
+		},
 	];
+
+	function getWorkerConfig(environmentName: string) {
+		return resolvedPluginConfig.type !== "assets-only"
```

Comment:
If we did keep the reduced config for preview mode then this change should work:

```suggestion
		return resolvedPluginConfig && resolvedPluginConfig.type !== "assets-only"
```

Reply:
Ah interesting. Now that I have refactored things I think you are right. We don't actually need there to be a worker config anywhere in preview mode, right?

The previous concern was the dealiasing of the virtual modules but now I strip the virtual prefix whether or not there is node.js compat turned on.

Reply:
Yep, I think if it's OK with you let's keep the reduced preview config. We can always reevaluate later in a separate PR.